### PR TITLE
LUCENE-9963 Add tests for alternate path failures in FlattenGraphFilter

### PR DESCRIPTION
**Pull request by glawson0**
# Description

Add tests for alternate path failures in FlattenGraphFilter.

Currently failing tests marked AwaitingFix so they will be ignored. Adding these tests before fix to allow others to also develop towards a fix.

# Solution

None

# Tests

Tests cover 3 problems that can arise for holes in alternate paths.
1) hole at start of path causes long path to not flatten correctly
2) hole at end of path causes long path to not move it's output point correctly
3) token stream ends before last shows up to end alternate path.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

### CODE REVIEW

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java [hunk 1]
```diff
@@ -314,5 +314,116 @@ public void testTwoLongParallelPaths() throws Exception {
...
+  public void testAltPathFirstStepHole() throws Exception {
+    TokenStream in =
+        new CannedTokenStream(
+            0,
+            3,
+            new Token[] {token("abc", 1, 3, 0, 3), token("b", 1, 1, 1, 2), token("c", 1, 1, 2, 3)});
+
+    TokenStream out = new FlattenGraphFilter(in);
+
+    assertTokenStreamContents(
```
Comment (msokolov) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java:328]: I'm curious to know what happens currently - how does the test fail? Could you maybe add a comment saying what the bad results are in each of these cases? Hmm I see you did in some cases below.

This stuff is crazy? TBH I am not all that familiar with how FlattenGraphFilter works, but I wonder if we could express some invariants cleanly and write a randomized test? There seem to be a lot of different edge cases.

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java [hunk 2]
```diff
@@ -314,5 +314,116 @@ public void testTwoLongParallelPaths() throws Exception {
...
+        out,
+        new String[] {"abc", "b", "c"},
+        new int[] {0, 1, 2},
+        new int[] {3, 2, 3},
+        new int[] {1, 1, 1},
+        new int[] {3, 1, 1},
+        3);
+  }
+  // Last node in an alt path releases the long path. but it doesn't exist in this graph
```
Comment (msokolov) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java:337]: nit: add a blank line

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java (hunk 1 above)
Comment (glawson0) [reply lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java:328]: A randomized test should be possible. We could build a random valid graph and try to flatten it. Checking offsets/length/increment would be difficult because the FlattenGraphFilter does change those as the graph gets "sausage-ized". You would have to know what tokens end up as edges between the same nodes in the final graph and what order they'ed be in.

An easy way to do this might be to start with a closed set of tokens, create a synonym dictionary dictionary with some random relations, pick some of the tokens to be stop words, add FlattenGraphFilter at the end, then generate some strings from your tokens and send them through.

Comment (glawson0) [reply lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java:339]: I've added a test that builds a graph then deletes some tokens. It already found another failure case in my fix.

Comment (mikemccand) [reply lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java:339]: Phew, thanks @glawson0!  This is super tricky.  I wish we could inline little `.png` pictures of the graphs in each case.  Maybe `graphviz`/`dot` supports ascii art output format?

Comment (glawson0) [reply lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java:339]: I couldn't find any tools within Lucene for visualizing `dot` graphs. There are a few tests that print out `dot` graphs when `tests.verbose=true`. I assume the idea is for the user to invoke graphviz themselves. We could continue with this pattern with the random and specific case tests. Is that what you're thinking about?

Comment (mikemccand) [reply lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java:339]: Oh sorry I missed this -- yeah I was thinking we could manually invoke graphviz/dot, and perhaps it supports ascii art output, and then copy/paste manually that ascii art as a comment in our java sources.  Seems like a long shot though!

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java [hunk 3]
```diff
@@ -314,5 +323,221 @@ public void testTwoLongParallelPaths() throws Exception {
...
+        out,
+        new String[] {"abc", "a", "b"},
+        new int[] {0, 0, 1},
+        new int[] {1, 1, 2},
+        new int[] {1, 0, 1},
+        new int[] {1, 1, 1},
+        2);
+  }
+
+  private CharsRef buildMultiTokenCarRef(
```
Comment (mikemccand) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java:446]: Hmm the method name is mis-spelled -- `CharsRef` suffix instead of `CarRef`?

**Review by mikemccand — APPROVED** [review]
I left minor typo comment but I think this is great that you pulled out (tricky!) tests ahead of the PR to fix it.  Thanks @glawson0.

**Comment by glawson0** [comment]
Thanks for the comments @mikemccand. The branch this PR is built in has multiple commits. For clarity about the contents of the commit message should I squash these down before it gets merged or should I leave squashing commits to the person doing the merging?

**Comment by mikemccand** [comment]
> The branch this PR is built in has multiple commits. For clarity about the contents of the commit message should I squash these down before it gets merged or should I leave squashing commits to the person doing the merging?

No need for you to pre-squash -- we can do that when merging.  GitHub gives us this convenient "Squash and Merge" (by default) button for just that.

I'll merge these added test cases soon.  Thanks @gautamworah96!

**Comment by mikemccand** [comment]
OK I merged and backported to 8.x!  Thanks @glawson0!  Now let's try to also finish #157.
